### PR TITLE
add feature: read grid from file

### PR DESCRIPTION
**Pull request by Hao2747**

### CODE REVIEW

File: main
Comment (kzhu2) [inline main:1]: Do not commit binary files

File: main.cpp [hunk 1]
```diff
@@ -184,20 +187,39 @@ class Grid {
     }
 };
 
+int main() {
+    std::ifstream inputFile("testbench/simple-50.txt"); // open file for input
```
Comment (kzhu2) [inline main.cpp:191]: Can you make sure that the files are included in the PR? I can see the files on the branch, but not in the PR. Not sure if its just a folder thing

File: main.cpp [hunk 2]
```diff
@@ -184,20 +187,39 @@ class Grid {
...
+                    idx++;
+                }
+            }
+            
+            all_grids.emplace_back(Grid(grid_size, matrix));
+        }
+        inputFile.close(); // close the file
+    }
+    else {
+        std::cerr << "Unable to open file" << std::endl;
```
Comment (kzhu2) [inline main.cpp:215]: Probably want to exit out/return here

File: main.cpp [hunk 3]
```diff
@@ -184,20 +187,39 @@ class Grid {
...
 
+int main() {
+    std::ifstream inputFile("testbench/simple-50.txt"); // open file for input
+    std::string line;
+    std::vector<Grid> all_grids;
+    dtype grid_size = 9;
+
+    if (inputFile.is_open()) { // check if file is open
+        while (std::getline(inputFile, line)) { // read lines until end of file
+            std::vector<std::vector<int>> matrix(9, std::vector<int>(9, -1));
```
Comment (kzhu2) [inline main.cpp:198]: Make more general by sqrt (num_chars on line) instead of hard coding 9

File: main.cpp [hunk 4]
```diff
@@ -184,20 +187,39 @@ class Grid {
...
+
+            int idx = 0;
```
Comment (kzhu2) [inline main.cpp:200]: Use size_t

File: main.cpp (hunk 3 above)
Comment (kzhu2) [inline main.cpp:198]: Change int to dtype

**Review by kzhu2 — CHANGES_REQUESTED** [review]
Good job. Just wanted to get your feedback and intentions on the future of getting the input. 
What will happen going forward in terms of input/output getting? Are there plans on adding  additional files to check the grid vs the actual reference (file to file compare) or just plan to use validate()?

**Comment by Hao2747** [comment]
> 

Great point! I would imagine file to file check would be faster? Also, it is always possible for validate() to be wrong. It is good to have ground truth. We could name problem set as simple-50-init.txt and solution as simple-50-ref.txt. So in command line, you put in simple-50 as the test name, it would read from simple-50-init.txt and compare it to simple-50-ref.txt. How do you think?

**Comment by kzhu2** [comment]
Sounds good and I agree. Just make sure its consistent so that when we add multiple files required for timing, its easy to switch between files inside the function.

The other functions like validate will still see use in case we need to do checks inside the some of the solvers.

**Pull request closed without merging**